hngprojects/hng_boilerplate_nestjs
# feat: Creating HelpCenter topics  by superadmin

---

## What does the PR do?

This PR introduces a SuperAdmin guard to the endpoint for creating HelpCenter topics. The guard ensures that only users with the SuperAdmin role can create new topics. This includes:

- Verifying that the token provided is valid and has the SuperAdmin role.
- Returning a 401 status code for missing or invalid authorization headers.
- Returning a 403 status code if the token does not have the SuperAdmin role.
- Correctly validating input fields and responding with a 400 status code for invalid requests.
- Ensuring that if a topic with the same title already exists, it returns a 400 status code with an appropriate message.
- If the user is not found, it returns a 404 status code.
- On successful creation of a help center topic, it returns a 201 status code with the appropriate response body.

## How should this be manually tested?

1. Ensure you have a valid SuperAdmin token.
2. Send a POST request to the following endpoint:

   ```
   POST /api/v1/help-center/topics
   ```

3. Include the Authorization header:

   ```
   Authorization: Bearer <access_token>
   ```

4. Provide a valid request body with the new help center topic details:

   ```json
   {
     "title": "New Help Center Topic",
     "content": "Detailed content for the new help center topic."
   }
   ```

## Success Response

On successful execution, you will receive the following response structure:

### help_center_create_success

```json
{
  "status_code": 201,
  "message": "Request successful",
  "data": {
    "id": "unique-topic-id",
    "title": "New Help Center Topic",
    "content": "Detailed content for the new help center topic.",
    "author": "First Last",
    "created_at": "2024-08-05T19:03:52.194Z"
  }
}
```

**Example Response Image:**
![create-help-center-success](https://github.com/user-attachments/assets/1b32b8bb-90e6-43dc-8ae5-e4374582150a)

## Unauthorized Response

If the token is missing or invalid, you will receive the following response:

### help_center_unauthorized

```json
{
  "status_code": 401,
  "message": "Unauthorized. Please log in."
}
```

## Forbidden Response

If the token does not have the SuperAdmin role, you will receive the following response:

### help_center_forbidden

```json
{
  "status_code": 403,
  "message": "Forbidden. You do not have the required permissions."
}
```

## Bad Request Response

If a topic with the same title already exists, you will receive the following response:

### help_center_bad_request

```json
{
  "status_code": 400,
  "message": "This question already exists."
}
```

## "Not Found" Response

If the user is not found, you will receive the following response:

### help_center_not_found

```json
{
  "status_code": 404,
  "message": "User not found."
}
```

## "Internal Server Error" Response

For unexpected server issues, you will receive the following response:

### help_center_internal_server_error

```json
{
  "status_code": 500,
  "message": "Internal server error. Failed to create the help center topic."
}
```

## Task

- Implemented SuperAdmin guard to restrict the creation of help center topics.
- Added method in `HelpCenterService` to handle topic creation with SuperAdmin restriction.
- Validated token and user role to ensure only SuperAdmins can create topics.
- Implemented appropriate error handling for invalid tokens, request bodies, duplicate topics, and non-existent users.
- Added unit tests for the endpoint covering authorization, input validation, and success cases.
- Updated API documentation to reflect the new endpoint and its requirements.

## Unit Test Evidence

**help_center_create_test**
![create-help-center-unit-test](https://github.com/user-attachments/assets/af9851aa-d219-4d67-9333-2538de450c21)

Reference Issue: [Link to Issue](https://linear.app/team-bingo/issue/BAC2-43/creating-a-help-article)
